workflow.go

```diff
@@ -0,0 +1,257 @@
+		err = workflow.ExecuteActivity(ctx, a.SendWelcomeEmail, s).Get(ctx, nil)
+		if err != nil {
+			logger.Error("Welcome Email Activity failed.", "Error", err)
+			return err
+		}
+
+		// Start trial period timer, async
+		trialDuration := s.TrialPeriodExpiration.Sub(workflow.Now(ctx))
+		trialTimer := workflow.NewTimer(ctx, trialDuration)
+
+		var timerError error
+		trialOver := false
+		selector.AddFuture(trialTimer, func(f workflow.Future) {
+			logger.Info("Trial has ended. Sending 'trial is over' email, then entering normal billing cycle.")
+			trialOver = true
+			timerError = f.Get(ctx, nil)
+		})
+
+		for !trialOver {
+			selector.
+				// Signal Handler for updating billing address or other customer info
+				AddReceive(workflow.GetSignalChannel(ctx, "update"), func(c workflow.ReceiveChannel, _ bool) {
+					var updateInfo UpdateSignal
+					c.Receive(ctx, &updateInfo)
+					logger.Info("Received update signal", "data", updateInfo)
+					s = &updateInfo.Subscription
+				}).
+				// Signal handler for canceling the subscription
+				AddReceive(workflow.GetSignalChannel(ctx, "cancel"), func(c workflow.ReceiveChannel, _ bool) {
+					var cancel UpdateSignal
+					c.Receive(ctx, &cancel)
+					logger.Info("Received cancel signal", "data", cancel)
+
+					cancelSub = cancel.CancelSubscription
+				})
+
+			err = workflow.Await(ctx, func() bool {
+				logger.Info("Current history length", "History Length", info.GetCurrentHistoryLength())
+				return info.GetCurrentHistoryLength() > EventHistoryThreshold || selector.HasPending()
+			})
+			if err != nil {
+				logger.Error("Awaiting history or selector failed.", "Error", err)
+				return err
+			}
+
+			if info.GetCurrentHistoryLength() > EventHistoryThreshold {
+				canError := workflow.NewContinueAsNewError(ctx, SubscriptionWorkflow, *s)
+				return canError
+			}
+
+			selector.Select(ctx)
+
+			if timerError != nil {
+				logger.Error("Trial timer failed.", "Error", err)
+				return err
+			}
+
+			if cancelSub {
+				logger.Info("Received request to cancel subscription. Not continuing into normal billing cycle. Workflow done.")
+				return nil
+			}
+		}
+
+		s.TrialPeriodExpiration = time.Time{}
+		err = workflow.ExecuteActivity(ctx, a.SendTrialExpiredEmail, s).Get(ctx, nil)
+		if err != nil {
+			logger.Error("SendTrialExpiredEmail Activity failed.", "Error", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *SubscriptionInfo) billingCycle(ctx workflow.Context) (err error) {
+	logger := workflow.GetLogger(ctx)
+	selector := workflow.NewSelector(ctx)
+	info := workflow.GetInfo(ctx)
+
+	var a *Activities
+	cancelSub := false
+
+	for !cancelSub {
+		// start of billing cycle, charge a subscription
+		err = workflow.ExecuteActivity(ctx, a.ChargeSubscription, s).Get(ctx, nil)
+		if err != nil {
+			logger.Error("ChargeSubscription Activity failed.", "Error", err)
+			return err
+		}
+
+		timerDuration := s.BillingPeriodDuration
```

Comment:
Durations in loops are best added to the last duration (and initial duration is often best as a fixed point, not related to workflow.Now which is when the workflow is running, not necessarily the date the action applied). For example, what if my billing duration is 5 days, but `ChargeSubscription` has some issue and takes a day? Then my billing duration is effectively 6 days.